File: backend/accounts/services/entitlements.py
from __future__ import annotations

from copy import deepcopy
from typing import Any

from accounts.models import SubscriptionTier

FeatureKey = str
LimitKey = str

DEFAULT_TIER = SubscriptionTier.FREE

ENTITLEMENT_CONFIG: dict[str, dict[str, dict[str, Any]]] = {
    SubscriptionTier.FREE: {
        "features": {
            "can_use_verified_badge": False,
            "can_use_priority_ranking": False,
        },
        "limits": {
            # Mirrors the current per-section card limit. This config is not enforced yet.
            "max_active_cards": 3,
            "monthly_boosts": 0,
        },
    },
    SubscriptionTier.PREMIUM: {
        "features": {
            "can_use_verified_badge": True,
            "can_use_priority_ranking": True,
        },
        "limits": {
            "max_active_cards": 10,
            "monthly_boosts": 5,
        },
    },
}


def get_user_tier(user) -> str:
    tier = str(getattr(user, "subscription_tier", None) or DEFAULT_TIER).strip()
    if tier in ENTITLEMENT_CONFIG:
        return tier
    return DEFAULT_TIER


def is_premium(user) -> bool:
    return get_user_tier(user) == SubscriptionTier.PREMIUM


def get_entitlements_for_user(user) -> dict[str, Any]:
    tier = get_user_tier(user)
    config = deepcopy(ENTITLEMENT_CONFIG[tier])

    return {
        "tier": tier,
        "is_premium": tier == SubscriptionTier.PREMIUM,
        "features": config["features"],
        "limits": config["limits"],
    }


def has_feature(user, feature_key: FeatureKey) -> bool:
    entitlements = get_entitlements_for_user(user)
    return bool(entitlements["features"].get(feature_key, False))


def get_limit(user, limit_key: LimitKey, default: int | None = None) -> int | None:
    entitlements = get_entitlements_for_user(user)
    value = entitlements["limits"].get(limit_key, default)
    return value if type(value) is int else default

File: frontend/src/lib/entitlements.ts
import type {
  EntitlementFeatureKey,
  EntitlementLimitKey,
  SubscriptionTier,
  User,
  UserEntitlements,
} from '@/types';

const FREE_TIER: SubscriptionTier = 'free';
const PREMIUM_TIER: SubscriptionTier = 'premium';

export const DEFAULT_USER_ENTITLEMENTS: UserEntitlements = {
  tier: FREE_TIER,
  is_premium: false,
  features: {
    can_use_verified_badge: false,
    can_use_priority_ranking: false,
  },
  limits: {
    max_active_cards: 3,
    monthly_boosts: 0,
  },
};

function isKnownTier(value: unknown): value is SubscriptionTier {
  return value === FREE_TIER || value === PREMIUM_TIER;
}

export function getUserTier(user: Pick<User, 'subscription_tier'> | null | undefined): SubscriptionTier {
  return isKnownTier(user?.subscription_tier) ? user.subscription_tier : FREE_TIER;
}

export function getUserEntitlements(
  user: Pick<User, 'subscription_tier' | 'entitlements'> | null | undefined,
): UserEntitlements {
  const fallbackTier = getUserTier(user);
  const entitlements = user?.entitlements;

  if (!entitlements || !isKnownTier(entitlements.tier)) {
    return {
      ...DEFAULT_USER_ENTITLEMENTS,
      tier: fallbackTier,
      is_premium: fallbackTier === PREMIUM_TIER,
      features: { ...DEFAULT_USER_ENTITLEMENTS.features },
      limits: { ...DEFAULT_USER_ENTITLEMENTS.limits },
    };
  }

  return {
    tier: fallbackTier,
    is_premium: fallbackTier === PREMIUM_TIER,
    features: {
      ...DEFAULT_USER_ENTITLEMENTS.features,
      ...(entitlements.features || {}),
    },
    limits: {
      ...DEFAULT_USER_ENTITLEMENTS.limits,
      ...(entitlements.limits || {}),
    },
  };
}

export function isPremiumUser(
  user: Pick<User, 'subscription_tier' | 'entitlements'> | null | undefined,
): boolean {
  return getUserEntitlements(user).is_premium;
}

export function hasEntitlementFeature(
  user: Pick<User, 'subscription_tier' | 'entitlements'> | null | undefined,
  featureKey: EntitlementFeatureKey,
): boolean {
  return getUserEntitlements(user).features[featureKey] === true;
}

export function getEntitlementLimit(
  user: Pick<User, 'subscription_tier' | 'entitlements'> | null | undefined,
  limitKey: EntitlementLimitKey,
  fallback = 0,
): number {
  const value = getUserEntitlements(user).limits[limitKey];
  return Number.isFinite(value) ? value : fallback;
}
